govcd/metadata_v2.go

```diff
@@ -0,0 +1,757 @@
+	MergeMetadataWithMetadataValuesAsync(metadata map[string]types.MetadataValue) (Task, error)
+	DeleteMetadataEntryAsync(key string) (Task, error)
+}
+
+// getMetadata is a generic function to retrieve metadata from VCD
+func getMetadata(client *Client, requestUri string) (*types.Metadata, error) {
+	metadata := &types.Metadata{}
+
+	_, err := client.ExecuteRequest(requestUri+"/metadata/", http.MethodGet, types.MimeMetaData, "error retrieving metadata: %s", nil, metadata)
+	return metadata, err
+}
+
+// addMetadata adds metadata to an entity.
+// If the metadata entry is of the SYSTEM domain (isSystem=true), one can set different types of Visibility:
+// types.MetadataReadOnlyVisibility, types.MetadataHiddenVisibility but NOT types.MetadataReadWriteVisibility.
+// If the metadata entry is of the GENERAL domain (isSystem=false), visibility is always types.MetadataReadWriteVisibility.
+// In terms of typedValues, that must be one of:
+// types.MetadataStringValue, types.MetadataNumberValue, types.MetadataDateTimeValue and types.MetadataBooleanValue.
+func addMetadata(client *Client, requestUri, key, value, typedValue, visibility string, isSystem bool) (Task, error) {
+
+	if isSystem && visibility == types.MetadataReadWriteVisibility {
+		return Task{}, fmt.Errorf("visibility cannot be %s when domain is SYSTEM", types.MetadataReadWriteVisibility)
+	}
+
+	apiEndpoint := urlParseRequestURI(requestUri)
+	newMetadata := &types.MetadataValue{
+		Xmlns: types.XMLNamespaceVCloud,
+		Xsi:   types.XMLNamespaceXSI,
+		TypedValue: &types.MetadataTypedValue{
+			XsiType: typedValue,
+			Value:   value,
+		},
+		Domain: &types.MetadataDomainTag{
+			Visibility: visibility,
+			Domain:     "SYSTEM",
+		},
+	}
+
+	if isSystem {
+		apiEndpoint.Path += "/metadata/SYSTEM/" + key
+	} else {
+		apiEndpoint.Path += "/metadata/" + key
+		newMetadata.Domain.Domain = "GENERAL"
+		if visibility != types.MetadataReadWriteVisibility {
+			newMetadata.Domain.Visibility = types.MetadataReadWriteVisibility
+		}
+	}
+
+	return client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodPut, types.MimeMetaDataValue, "error adding metadata: %s", newMetadata)
+}
+
+// addMetadataAndWait adds metadata to an entity and waits for the task completion.
+// The function supports passing a value that requires a typed value that must be one of:
+// types.MetadataStringValue, types.MetadataNumberValue, types.MetadataDateTimeValue and types.MetadataBooleanValue.
+// Visibility also needs to be one of: types.MetadataReadOnlyVisibility, types.MetadataHiddenVisibility or types.MetadataReadWriteVisibility
+func addMetadataAndWait(receiver metadataAsync, key, value, typedValue, visibility string, isSystem bool) error {
+	task, err := receiver.AddMetadataEntryWithVisibilityAsync(key, value, typedValue, visibility, isSystem)
+	if err != nil {
+		return err
+	}
+
+	return task.WaitTaskCompletion()
```

Comment:
Note to reviewers: Original `metadata.go` had a `Refresh()` call on the receiver objects. I've removed them as it didn't make much sense (metadata is handled outside the entities).
Let me know if I'm taking something for granted.